### ISSUE TITLE
Cache package updates

### ISSUE THREAD

**Issue by andrewdavidwong** [QubesOS/qubes-issues#1957]
It's common for users to have multiple TemplateVMs that download many of the same packages when being individually updated. Caching these packages (e.g., in the UpdateVM) would allow us to download a package only once, then make it available to all the TemplateVMs which need it (and perhaps even to dom0), thereby saving bandwidth.

This has come up on the mailing lists several times over the years:
* https://groups.google.com/d/topic/qubes-devel/bPYvm18e8z0/discussion
* https://groups.google.com/d/msg/qubes-devel/Qu89S2CdRMU/f0S9VHix6kIJ
* https://groups.google.com/d/msg/qubes-devel/cekPfBqQMOI/uppr45K8AwAJ

Here's a blog post about setting up a squid caching proxy for DNF updates on baremetal Fedora:
* http://www.hergert.me/blog/2015/07/29/caching-updates.html

**Comment by taradiddles** [comment]
It's indeed a common problem when deploying fedora vms/containers, or with server farms. Debian has apt-cacher(ng) but fedora doesn't have something similar.

Solutions that came up:

- squid as a caching proxy (as noted by andrew).
Need to check cache expiry time, max object size, ..., in order to minimize cache miss.
Similar notes:
https://www.berrange.com/posts/2015/12/09/setting-up-a-local-caching-proxy-for-fedora-yum-repositories/

- pulp ; http://www.pulpproject.org/
too heavy ?

- cobbler ; https://fedorahosted.org/cobbler/
From the doc: "Cobbler can also optionally help with [...] yum package mirroring infrastructure"
But - looks specific to fedora and maybe too heavy too.

- guru labs automatic mirror - https://www.gurulabs.com/goodies/guru-guides/YUM-automatic-local-mirror/
Needs a web server, and without tweaks to the yum repo files, to be authoritative for well-known fedora mirro domains. Other than that, looks like a rather "clean" solution.

- mrepo ; http://dag.wiee.rs/home-made/mrepo/
Oldish but may get the work done.

Anyway, instead of having specific tools for each distro it would be wiser to have a generic solution.
So - all in all, the squid solution may be the best one, with cache misses rate being something to investigate.

**Comment by marmarek** [comment]
Actually apt-cacher works for Fedora too :)
Maybe we can simply use it instead of tinyproxy as update proxy?

**Comment by taradiddles** [comment]
apt-cacher-ng works on fedora for mirroring debian stuff, but does it really work for mirroring (d)rpms/metadata downloaded with yum/dnf ?

From the doc [1]:  "6.3 Fedora Core - Attempts to add apt-cacher-ng support ended up in pain and the author lost any motivation in further research on this subject. "

[1] https://www.unix-ag.uni-kl.de/~bloch/acng/html/distinstructions.html#hints-fccore

**Comment by marmarek** [comment]
Yes, I've seen this. But in practice it works. The only problem is
dynamic mirror selection - it may make caching difficult (when each time
different mirror is selected).

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by adrelanos** [comment]
Marek Marczykowski-Górecki:
> Actually apt-cacher works for Fedora too :)
> Maybe we can simply use it instead of tinyproxy as update proxy?

Can it also let through non-apt traffic? Specifically I am wondering
about tb-updater.

**Comment by marmarek** [comment]
> Can it also let through non-apt traffic? Specifically I am wondering
> about tb-updater.

That's interesting question - if you have apt-cacher-ng instance handy,
it worth a try. Anyway it has quite flexible configuration, so probably
doable.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by adrelanos** [comment]
I don't think there is a generic solution that works at the same time
well enough for both, Debian and Fedora based. Why do we need a generic
all at once solution anyhow? Here is what I suggest:

- Let's keep tinyproxy as is. As fallback. And for misc traffic.
(tb-updater, user custom stuff and what not.)

- Let's install apt-cacher-ng and the fedora caching proxy by default in
the UpdateVM.

- Let's configure Debian based VMs to use apt-cacher-ng.

- Let's configure Fedora based VMs to use the fedora caching proxy.

What do you think?

@marmarek
>> Can it also let through non-apt traffic? Specifically I am wondering
>> about tb-updater.
> 
> That's interesting question - if you have apt-cacher-ng instance handy,
> it worth a try. Anyway it has quite flexible configuration, so probably
> doable.

I've read all the config, and tired, does not seem possible but never
mind as per my above suggestion.

**Comment by marmarek** [comment]
It will require more resources (memory), somehow wasted when one use for example only Debian templates. But maybe it is possible to activate those services on demand (socket activation comes to my mind). It will be even easier for qrexec-based updates proxy.

**Comment by taradiddles** [comment]
@adrelanos 

> Why do we need a generic all at once solution anyhow

I'm all for 100% caching success rate with a specific mechanism for each distro, but do Qubes developpers/contributors have time to develop/support that feature ?
If yes, that's cool ; otherwise, a solution like squid would be easy to implement, and since it's distro agnostic it will help not only the supported distros (fedora, debian, arch?), but also other distributions that users install in HVMs (even windows then). The problems/unknowns with squid are the cache miss rate, the cache disk usage in order to minimize those, and the use of different mirrors with yum (although I find out that I usually always connect to the same one).

**Comment by qjoo** [comment]
I'm using polipo proxy =>  tor to cache updates. I also modified the repo configuration to use one specific update server instead of dynamically selecting it. I'm planing to document my setup and will post a link here.

**Comment by kalkin** [comment]
Just wanted to throw in https://github.com/yevmel/squid-rpm-cache I planned to setup a dedicated squid vm and use the above mentioned config/plugin to cache rpms, but never found the time for it. 

> The problems/unknowns with squid are the cache miss rate, the cache disk usage in order to minimize those, and the use of different mirrors with yum (although I find out that I usually always connect to the same one).

Currently i just use my NAS which has a "normal" squid running as caching proxy. I have an ansible script which generates me my templates. In the templates I replaced the `metalink` parameter with `baseurl` to the nearest Fedora mirror, in `/etc/yum.repos.d/fedora.repo`. In `/etc/yum.conf` I replaced the proxy option with my NAS proxy and allowed TempalteVMs to connect to it.

**Comment by marmarek** [comment]
My experience with squid is horrible in terms of resources (RAM, I/O usage) for small setups. Looks like an overkill for just downloading updates from a few templates from time to time.

**Comment by andrewdavidwong** [comment]
It doesn't have to be squid. Is there a better option?

**Comment by andrewdavidwong** [comment]
(Never mind. Just re-read the thread.)

**Comment by adrelanos** [comment]
I don't like saying this, but we should also consider making this an additional, non-default option or wontfix also. I like apt-cacher-ng very much and use it myself. However, introducing it by default into Qubes would lead to [new issues](https://bugs.debian.org/cgi-bin/pkgreport.cgi?package=apt-cacher-ng), more users having issues with upgrading due to added technical complexity. There are corner cases where apt-cacher-ng introduces new issues, such as showing `Hash Sum mismatch` errors during `apt-get update`.

**Comment by taradiddles** [comment]
@marmarek

FWIW I have squid installed on an embedded router (RB450g) for a 25+ people office and it's been running for literally ages without any problem. There's a strict bandwidth control (delay pools), which is usually the biggest offender in terms of resources, but squid's memory usage has constantly been < 20 Mo and highest CPU usage < 6%. Granted, the office's uplink speed is low - in the megabits/s range - but the resources available for updateVM are in another league compared to the embedded stuff and the setup - only caching - is not fancy.

tl;dr, squid is not as bad as it used to be years ago.

@adrelanos 

The issues you mention reinforce my concern that it will be too time-consuming for Qubes devs to support distro-specific solutions. A simple generic one, even if not optimal is still better than nothing at all, rather than "wontfix".
Plus, users kalkin and qjoo seem to have a working solution, why not try those ?

just my 2c  - not pushing for anything, you guys are doing a great work !

**Comment by andrewdavidwong** [comment]
At the very least, we should provide some documentation (or suggestions or pointers in the documentation) regarding something like @taradiddles's router solution. Qubes users are more likely than the average Linux user to have multiple machines (in this case, virtual) downloading exactly the same updates.

**Comment by Rudd-O** [comment]
Looks like what you want is Squid with an adaptive disk cache size (for storing packages in the volatile `/var/cache/squid` directory), and configured with no memory cache.  Since the config file can be in a different place and the unit file can be overridden to specify the Qubes specific config file, it may work very well for this purpose.  Squid is goddamn good these days, and it supports regex-based filters (plus you can block methods other than GET, and you can support proxy caching FTP sites).

OTOH, it's always a security footprint issue to run a larger codebase for a cache.  Also, Squid caching can be ineffective if multiple VMs download files from different mirrors (remember that the decision of which mirror to use is left practically at random to the VM calling onto the Squid proxy to do its job).

For those reasons, it may be wise to investigate solutions that do a better job of proxy caching using a content-addressable store, or matching file names.

**Comment by Rudd-O** [comment]
Perhaps a custom Go-based (to prevent security vulns) cache that can listen for requests using the net/http module, and proxy them to the VMs?  This has potential to be a very efficient solution too, as a Go program would have a minuscule memory footprint.

**Comment by kalkin** [comment]
@Rudd-O Have a look at this

**Comment by Rudd-O** [comment]
Looking.  Note we need something like that for Debian as well.

**Comment by Rudd-O** [comment]
The code is not idiomatic Go and there are some warts there that I would fix before including it anywhere.  Just as a small example on https://github.com/mojaves/yumreproxyd/blob/master/yumreproxy/yumreproxy.go#L33 you can see he is using a nil value as a sort of a bool.  That is not correct -- the return type should be (bool, struct).

**Comment by Rudd-O** [comment]
https://github.com/mojaves/yumreproxyd/blob/master/yumreproxy/yumreproxy.go#L73 <- also problematic.  `TODO: path sanitization` is not what you want in secure software.

But the BIGGEST problem, is that the program appears not to give a shit about concurrency.  Save into cache and serve from cache can have a race, and no locking is performed, nor are channels being used there.  Big fat red flag.  The right way to do that by communicating with the Cache aspect of the application through channels -- send request to the Cache, await for response, if not available, then download file, send storage to the Cache, await for response.

Also, all content types returned are application/rpm.  That's wrong in many cases.

BUT, that only means that project can be extended or rewritten, and it should not be very difficult to do so.

**Comment by rustybird** [comment]
I just uploaded the Squid-based https://github.com/rustybird/qubes-updates-cache (posted to qubes-devel too, but stuck in moderation or something)

**Comment by rustybird** [comment]
The latest commit (-57 lines, woo) reworks qubes-updates-cache to act as a drop-in replacement for qubes-updates-proxy. No changes to the client templates are needed at all now.

**Comment by marmarek** [comment]
How much memory does it use? I.e. is it a good idea to have it instead of tinyproxy by default, or give the user a choice?

**Comment by taradiddles** [comment]
FWIW I had a similar setup running after my last post, the difference being that I used/tweaked the store_id program mentioned by @kalkin in an earlier post [1]. But there were many cache misses ; a quick look at the log showed that different mirrors would send different mime types for the same rpm (or repo) file, so that might be the culprit. Other tasks piled up and I didn't have time to work on that.

@marmarek :  after boot, memory = ~30Mo (as far as you can trust ps). But I guess the question is more on the long term use, after squid has cached many objects. Rusty used 'cache_mem=0', so there shouldn't be a huge difference in mem usage, but he might have more statistics.

@rustybird : tinyproxy's configuration is quite locked down, maybe that would be a good idea to do the same with squid's ? I'm also not sure it is a good idea to mess with the cache ids for files other than rpm/repo (and deb/...).

For instance, stuff like:

```
acl localnet src 10.137.0.0/16
acl http_ports port 80
acl SSL_ports port 443
acl CONNECT method CONNECT
http_access deny to_localhost
http_access deny CONNECT !SSL_ports
http_access allow http_ports
http_access allow SSL_ports
http_access deny all

# that one was from https://github.com/yevmel/squid-rpm-cache
# have to understand why that's changed
#refresh_pattern .              0       20%     4320
#                 3 month               12 month
refresh_pattern . 129600        33%     525600

# cache only specific files types
acl rpm_files urlpath_regex \/Packages\/.*\.rpm
acl repodata_files urlpath_regex \/repodata\/.*\.(|sqlite\.xz|xml(\.[xg]z)?)
cache allow rpm_files
cache allow repodata_files
cache deny all

```

[1] https://github.com/yevmel/squid-rpm-cache

**Comment by rustybird** [comment]
@marmarek:
> How much memory does it use?

With `DefaultMemoryAccounting=yes` in /etc/systemd/system.conf, the following values were observed in /sys/fs/cgroup/memory/system.slice/qubes-updates-cache.service/memory.memsw.max_usage_in_bytes:

- Squid first started, created new cache dir = 41 MiB 
- Upgraded a new clone of qubes-template-fedora-23-3.0.4-201601120722 (~450 packages) = 202 MiB
- Upgraded another new clone of the same template, ~100% cache hits = still 202 MiB
- Squid restarted, uses filled cache dir = 16 MiB

That's already the latest commit, which sets `memory_pools off` in the Squid config to allow the system to reclaim unused memory. But apparently Squid doesn't free() aggressively enough yet for our purposes.

@taradiddles:
> But there were many cache misses ; a quick look at the log showed that different mirrors would send different mime types for the same rpm (or repo) file, so that might be the culprit.

Yes, that seems to happen sometimes, probably because .drpm is a relatively young file extension. Is it possible to make Squid ignore the MIME type header?

> tinyproxy's configuration is quite locked down, maybe that would be a good idea to do the same with squid's ?

Definitely. IIRC Whonix also wants some sort of magic string from the proxy port? Paging @adrelanos :)

> I'm also not sure it is a good idea to mess with the cache ids for files other than rpm/repo (and deb/...). 

So far I haven't seen the regexes in https://github.com/rustybird/qubes-updates-cache/blob/master/usr/lib/qubes/updates-cache-dedup#L6-L7 match anything else besides metadata and packages. Files aren't listed explicitly because that's such a hassle to maintain for all compression formats and package types, e.g. Debian source packages didn't work with qubes-updates-proxy when tinyproxy still used filters.

**Comment by rustybird** [comment]
> Definitely. IIRC Whonix also wants some sort of magic string from the proxy port? Paging @adrelanos :)

Sorry, never mind, I literally found something with `grep -r magic /etc/tinyproxy`. Will check that out.

**Comment by adrelanos** [comment]
Ivan:
> tinyproxy's configuration is quite locked down, maybe that would be a good idea to do the same with squid's ?

Tinyproxy configuration was relaxed some time ago. There was a ticket
and discussion. In short: locking down tinyproxy does not improve actual
security. Users who explicitly configure their applications to use the
updates proxy should be free to do so.

**Comment by rustybird** [comment]
@adrelanos:
> Tinyproxy configuration was relaxed some time ago. There was a ticket and discussion. In short: locking down tinyproxy does not improve actual security. Users who explicitly configure their applications to use the updates proxy should be free to do so.

There's the "Squid Manager" though, which I've restricted access to in commit https://github.com/rustybird/qubes-updates-cache/commit/0da1dcd222263af73f6b52ac6d2b5d07444474a8 -- along with a basic sanity check that requests are coming from 10.137.*.

Also a paragraph on how to use qubes-updates-manager with Whonix at the moment: https://github.com/rustybird/qubes-updates-cache/blob/3b9d5e153f89b551e9b38f82928cbc7c9c2f7ba3/README#L32-L35 (works nicely BTW, tons of cache hits across Debian / Whonix GW / Whonix WS)

**Comment by adrelanos** [comment]
I have just now finished documenting Qubes-Whonix torified updates proxy:
https://www.whonix.org/wiki/Dev/Qubes#Torified_Updates_Proxy

In essence, Whonix TemplateVMs are getting the output of `UWT_DEV_PASSTHROUGH="1" curl --silent --connect-timeout 10 "http://10.137.255.254:8082/"` and grep it for `<meta name="application-name" content="tor proxy"/>`. If that matches, that test is considered successful.

Of course qubes-updates-cache's squid should only include the magic string, if it is actually torrified, i.e. run inside sys-whonix.

Do you know if it is possible to conditionally inject this magic string? Or if not, we need to modify Qubes-Whonix torified updates check to do something supported by squid.

I am wondering if any whonix-gw-firewall modifications will be required. Current tinyproxy rules:
https://github.com/Whonix/whonix-gw-firewall/blob/724a0fc0546c83555a008cd1b7b03c048519121a/usr/bin/whonix_firewall#L310-L328

Does squid support outgoing proxy settings? Can squid be configured to use a Tor SocksPort?

**Comment by rustybird** [comment]
> Does squid support outgoing proxy settings? Can squid be configured to use a Tor SocksPort?

Haven't found anything about outgoing HTTP proxies. [Semi-official Socks support can be added in during compilation via libsocks](http://wiki.squid-cache.org/Features/Socks), which Debian doesn't seem to do, but ...

> I am wondering if any whonix-gw-firewall modifications will be required. Current tinyproxy rules:
> https://github.com/Whonix/whonix-gw-firewall/blob/724a0fc0546c83555a008cd1b7b03c048519121a/usr/bin/whonix_firewall#L310-L328

... I think you'd only need to change `--uid-owner tinyproxy` to `--uid-owner squid`.

> AFAIR in case of tinyproxy it is placed in default error page. Squid should allow the same.

Yes, the relevant file to modify is /usr/share/squid-langpack/en/ERR_INVALID_URL from Debian package squid-langpack.

**Comment by rustybird** [comment]
All the security implications of using qubes-updates-cache I could think of:
https://github.com/rustybird/qubes-updates-cache/blob/master/README#L8

**Comment by rustybird** [comment]
Some news:

- Made qubes-update-cache work on Debian, incl. Whonix gateways pending PRs https://github.com/Whonix/whonix-gw-firewall/pull/1 and https://github.com/Whonix/qubes-whonix/pull/2
- Ordered the systemd service after bind-dirs.sh/rc.local and made it simpler and more reliable, which should fix TCP_SWAPFAIL_MISS cache corruption
- Switched to asynchronous cache backend, reducing memory consumption to ~ 130 MiB after one Fedora template upgrade. Still have to find out why adding new objects leaks memory at all
- Rewrote the URL rewriting script in pure bash with no child processes. Now in addition to dedup it also transparently upgrades some hosts from HTTP to HTTPS (www.whonix.org, deb.torproject.org, dl.google.com, mirrors.kernel.org)

**Comment by adrelanos** [comment]
@rustybird 

> Made qubes-updates-cache work on Debian, incl. Whonix gateways pending PRs Whonix/whonix-gw-firewall#1 and Whonix/qubes-whonix#2 

This is done btw.

**Comment by qjoo** [comment]
I use polipo as a caching proxy between template VMs and Tor SOCKS port. It has SOCKS support and might be more lightweight than squid?

**Comment by rustybird** [comment]
@adrelanos:

>> Made qubes-updates-cache work on Debian, incl. Whonix gateways pending PRs Whonix/whonix-gw-firewall#1 and Whonix/qubes-whonix#2

> This is done btw.

Looks like whonix-gw-firewall needs a version bump, and qubes-whonix 5.3-1 hasn't been uploaded yet?

@qjoo:

> I use polipo as a caching proxy between template VMs and Tor SOCKS port. It has SOCKS support and might be more lightweight than squid?

It doesn't seem to support deduplication or (transparent) rewriting of URLs :(

**Comment by adrelanos** [comment]
Rusty Bird:
> @adrelanos:
> 
>>> Made qubes-updates-cache work on Debian, incl. Whonix gateways pending PRs Whonix/whonix-gw-firewall#1 and Whonix/qubes-whonix#2
> 
>> This is done btw.
> 
> Looks like whonix-gw-firewall needs a version bump, and qubes-whonix 5.3-1 hasn't been uploaded yet?

Yes. The usual ETA to reach Whonix stable users is the next release,
Whonix 14.

**Comment by adrelanos** [comment]
* I'll release a qubes-whonix package with your qubes-updates-cache changes soon. (Currently in developers repository, contains some other fixes.)
* Should not writing to /etc i.e. [/etc/systemd/system/multi-user.target.wants/qubes-updates-cache.service](https://github.com/rustybird/qubes-updates-cache/blob/master/etc/systemd/system/multi-user.target.wants/qubes-updates-cache.service) better be avoided and standard distribution default systemd folders /lib/systemd/system be used?
* Where is the code for qubes-updates-cache.service?

**Comment by marmarek** [comment]
> Where is the code for qubes-updates-cache.service?

Here: https://github.com/rustybird/qubes-updates-cache/blob/master/usr/lib/systemd/system/qubes-updates-cache.service

> Should not writing to /etc i.e. /etc/systemd/system/multi-user.target.wants/qubes-updates-cache.service better be avoided and standard distribution default systemd folders /lib/systemd/system be used?

The standard way is to create such symlink in post-installation script (preferably using presets). But since the service is controlled by qvm-service, it may be indeed good idea to provide the symlink in the package. In such a case it should live in /lib/systemd/system and be relative one.

**Comment by rustybird** [comment]
Short update:

- The TCP_SWAPFAIL_MISS cache corruption still happens. It looks like Squid 3 just cannot deal with unclean shutdowns.

- Squid 4 fixes the memory leak! But the latest beta (4.0.12) is still too crashy to use.

@marmarek:

> But since the service is controlled by qvm-service, it may be indeed good idea to provide the symlink in the package. In such a case it should live in /lib/systemd/system and be relative one.

It's currently created in /etc, as if qubes-updates-cache.service was listed in https://github.com/QubesOS/qubes-core-agent-linux/blob/master/vm-systemd/75-qubes-vm.preset just like qubes-updates-proxy.service.

But I'll have to move at least the actual qubes-updates-cache.service to `$(pkg-config --variable=systemdsystemunitdir systemd)` anyway, installing it to /usr/lib/systemd/system is wrong for Debian. Then the symlink could be moved there, too.

**Comment by Rudd-O** [comment]
I would really like to urge folks into developing a custom cache solution using the very mature Go libraries that exist for HTTP and proxying.  It will be memory-safe (no pointer bullshit), it will be far smaller than trying to shoehorn Squid into this role, and it will be trivial to provide a proper solution that will cache requested file names based on content.

**Comment by adrelanos** [comment]
@rustybird:
>Looks like whonix-gw-firewall needs a version bump, and qubes-whonix 5.3-1 hasn't been uploaded yet?

@adrelanos:
> I'll release a qubes-whonix package with your qubes-updates-cache changes soon. (Currently in developers repository, contains some other fixes.)

It's in Whonix jessie (stable) repository for a few days now. (And if you [reinstall](https://www.qubes-os.org/doc/reinstall-template/) Qubes-Whonix 13 from qubes-tempaltes-community repository, it is also included.)

**Comment by rustybird** [comment]
The latest [qubes-updates-cache](https://github.com/rustybird/qubes-updates-cache) has many new rewriting rules that transparently upgrade repository URLs to [HTTPS](https://github.com/rustybird/qubes-updates-cache/blob/master/etc/qubes-updates-cache/urls/20-https.bash), and [optionally](https://github.com/rustybird/qubes-updates-cache/blob/25d1fa2ed2981bf1284ff4ca117b3b589bbdfbf7/README#L46-L51) to [.onion](https://github.com/rustybird/qubes-updates-cache/blob/master/etc/qubes-updates-cache/urls/10-onion.bash.EXAMPLE) (#2576).

Current coverage:

 Repository      |  HTTPS  | HTTP .onion
-----------------|---------|-------------
 yum.Qubes       | upgrade | upgrade
 deb.Qubes       | upgrade | upgrade
 Whonix          | upgrade | upgrade
 Debian          | upgrade | upgrade
 Debian Security | upgrade | upgrade
 Fedora          | upgrade |    -
 RPM Fusion      | upgrade |    -
 Tor Project     | upgrade | upgrade
 Google          | upgrade |    -
 Fedora-Cisco    | native  |    -
 Adobe           |    -    |    -

**Comment by tlaurion** [comment]
@adrelanos : I'm confused. Was qubes-updates-cache merged into Whonix 14?

**Comment by rustybird** [comment]
@tlaurion:

> Was qubes-updates-cache merged into Whonix 14?

No, installation is still manual. The Whonix PRs just made it so that if you install qubes-updates-cache in a whonix-gw VM, you don't have to fiddle with Whonix's firewall and the torification safety check.

**Comment by andrewdavidwong** [comment]
@rustybird, are you still working on this, or are you waiting for your PoC to be reviewed?

**Comment by rustybird** [comment]
@andrewdavidwong:
> are you still working on this, or are you waiting for your PoC to be reviewed?

It's still missing the Makefiles etc. for qubes-builder integration.

Sad to say, it looks like transparent caches are becoming less viable as more and more package repository definitions on the client VM templates switch to HTTPS. Previously, they made interceptable (= cacheable) HTTP requests and qubes-updates-cache could transparently upgrade to HTTPS on cache miss. Now it's opaque and uncacheable.

I don't know what to do about that. Reverting all repository definitions to HTTP is a non-starter, because that would harm users of the non-caching qubes-updates-proxy (i.e. tinyproxy, which AFAIK can't rewrite URLs to upgrade them to HTTPS) as well as users installing software from an AppVM/StandaloneVM without any updates proxy. So there would have to be a mechanism to modify repository definitions just for qubes-updates-cache users, or alternatively some cooperative TLS MITM contraption between templates and the cache. Both sound horrible...

**Comment by unman** [comment]
I have followed the first (horrible) option, using apt-cacher-ng.
I rewrite the definitions in templates to http://HTTPS///repo_name format, and apt-cacher-ng then transforms this to https requests. Horrible indeed, but works. 
Simply implemented with one salt call.

Of course, I need a script to rewrite the definitions in TemplateBased qubes.

**Comment by tlaurion** [comment]
The main solution replacing this one seems to live [here](https://github.com/unman/notes/tree/master/config/cacher) as recommended [here](https://groups.google.com/forum/#!msg/qubes-users/pH5B-2J4-qs/e11eBmghAAAJ)

**Comment by andrewdavidwong** [comment]
Ah, so @unman's caching proxy is replacing @rustybird's qubes-updates-cache as the contribution for this issue?

@unman, are you ready for your caching proxy to undergo review?

**Comment by tlaurion** [comment]
@unman : This work completion could be done under grant, depending on how good it is. It was considered horrible by unam himself [here](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-502108249).

**Comment by SvenSemmler** [comment]
For what it's worth:

- apt-cacher-ng does indeed work as a drop in replacement (as demonstrated and documented by @unman) 
- Once the few https:// references are changed to http://HTTPS/// in the TemplateVMs it just works at least with fedora, debian & ubuntu (haven't tested other templates)
- If this would be the default in Qubes and the Templates would come with http://HTTPS/ references ... the user would only hit issues when installing additional repositories (e.g. signal). But this could be described in documentation on qubes-os.org

Maybe make it an option during installation as @tlaurion propsed and have salt do the work? It might be "horrible" but it does work and offers great benefit to many users (many templates, low bandwidth, updating via TOR, etc.)

**Comment by tlaurion** [comment]
@unman @marmarek : So I understand that the apt-cacher-ng solution would stay hackish for the general user and implies user awareness of the caching proxy mechanism... Not desirable. Unless:
-  a wrapper in apt/dnf and other distros are created to modify the URLs contacted on the fly in case of apt-cacher-ng deployment of @unman  (not optimal)
- a hook on dom0 update-manager if apt-cacher-ng is deployed, calling @unman salt recipe to force modification of all TemplateVMs package manager's configuration files of all possibe distros to replace normal https references to the new one before attempting to update any package (better, but still not optimal IMO)

Else, I understand that @rustybird solution could be used by forcing all updates to go through http over tor and rely on package manager integrity checks on downloaded package. Also not optimal.

Is there any viable solution here? @marmarek ?

**Comment by tlaurion** [comment]
@marmarek : Questionning the actual flow of updating TemplateVMs.

If instead of spawning different dispvm for each TemplateVM, one was spawned to download the different updates for all TemplateVMs in a upgrade invocation, caching in /var the rpms, deb etc online, each subsequent TemplateVM downloading those same archives could hit a cache and download from that cache instead of online. No?

**Comment by tasket** [comment]
@tlaurion I'm probably missing something here, but it seems to me an underlying problem here is the whole "transparent caching proxy" concept.

It might be better to dedicate a VM to simply mirroring selected repositories, and that "repo VM" serving as the update VM. This probably has a disk space overhead, but seems like it would be a great opt-in to have, especially when it has the added benefit of concealing which packages are being deployed by the user.

**Comment by tlaurion** [comment]
@tasket @marmarek : Who needs "transparent caching proxy" if we can have a "disk qubes-updates-cache"?  That would still be a qubes-updates-cache IMO.

Right now, TemplateVM upgrades are launched one TemplateVM at a time.An underlying temp LVM/loop dev could be used to pass the updates cache from one TemplateVM to the other, which would validate the integrity of the packages prior of installing them? That would fill the need.  No?

Let me paraphrase current [template in-place upgrade instructions](https://www.qubes-os.org/doc/template/fedora/upgrade/#summary-instructions-for-standard-fedora-templatevms) which could be generalized as a qubes-updates-cache mechanism:

Prepare upgrade from dom0
```
[user@dom0 ~]$ truncate -s 5GB /var/tmp/template-upgrade-cache.img
[user@dom0 ~]$ dev=$(sudo losetup -f --show /var/tmp/template-upgrade-cache.img)
[user@dom0 ~]$ sudo mkfs.ext4 $dev
```
Do for each TemplateVM to use package cache:
```
[user@dom0 ~]$ qvm-block attach fedora-<new> dom0:${dev##*/}
[user@fedora-<new> ~]$ sudo mount /dev/xvdi /mnt/removable
[user@fedora-<new> ~]$ sudo dnf --releasever=<new> --setopt=cachedir=/mnt/removable --best --allowerasing distro-sync -y
[user@fedora-<new> ~]$ sudo halt
```
Cleanup dom0 after system upgrade session:
```
[user@dom0 ~]$ sudo losetup -d $dev
[user@dom0 ~]$ rm /var/tmp/template-upgrade-cache.img
```

**Comment by tlaurion** [comment]
@andrewdavidwong @mfc : bounty tage could be added to realize https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-687200236

**Comment by unman** [comment]
On Fri, Sep 04, 2020 at 07:56:47AM -0700, tlaurion wrote:
> @tasket @marmarek : Who needs "transparent caching proxy" if we can have a "disk qubes-updates-cache"?  That would still be a qubes-updates-cache IMO.
> 
> Right now, TemplateVM upgrades are launched one TemplateVM at a time.An underlying temp LVM/loop dev could be used to pass the updates cache from one TemplateVM to the other, which would validate the integrity of the packages prior of installing them? That would fill the need.  No?
> 
> Let me paraphrase current [template in-place upgrade instructions](https://www.qubes-os.org/doc/template/fedora/upgrade/#summary-instructions-for-standard-fedora-templatevms) which could be generalized as a qubes-updates-cache mechanism:
> 
> Prepare upgrade from dom0
> ```
> [user@dom0 ~]$ truncate -s 5GB /var/tmp/template-upgrade-cache.img
> [user@dom0 ~]$ dev=$(sudo losetup -f --show /var/tmp/template-upgrade-cache.img)
> [user@dom0 ~]$ sudo mkfs.ext4 $dev
> ```
> Do for each TemplateVM to use package cache:
> ```
> [user@dom0 ~]$ qvm-block attach fedora-<new> dom0:${dev##*/}
> [user@fedora-<new> ~]$ sudo mount /dev/xvdi /mnt/removable
> [user@fedora-<new> ~]$ sudo dnf --releasever=<new> --setopt=cachedir=/mnt/removable --best --allowerasing distro-sync -y
> [user@fedora-<new> ~]$ sudo halt
> ```
> Cleanup dom0 after system upgrade session:
> ```
> [user@dom0 ~]$ sudo losetup -d $dev
> [user@dom0 ~]$ rm /var/tmp/template-upgrade-cache.img
> ```
> 

Limiting to 1 VM template at a time? *very* sub-optimal.
There's a significant overhead too in maintenance terms, and cross
distro support.
The benefits of an existing caching proxy are that these issues have
already been resolved.

**Comment by andrewdavidwong** [comment]
> @andrewdavidwong @mfc : bounty tag could be added to realize [#1957 (comment)](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-687200236)

I don't understand. Did someone offer a bounty for this issue? I don't see any mention of one.

**Comment by tlaurion** [comment]
@andrewdavidwong : Insurgo collective could pay upon agreed price for deliverables and upon PoW.

**Comment by andrewdavidwong** [comment]
> @andrewdavidwong @mfc : bounty tag could be added to realize [#1957 (comment)](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-687200236)

> @andrewdavidwong : Insurgo collective could pay upon agreed price for deliverables and upon PoW.

`bounty` tag added per your request, but I think you'll need to provide more details (e.g., amount of bounty, exact work required to claim) before you get any bites.

**Comment by tlaurion** [comment]
@unman : thanks for your reply.
> Limiting to 1 VM template at a time? *very* sub-optimal. 

This is how Qubes Updates widget's launched upgrade process currently works: one TemplateVM being upgraded at a time.
> There's a significant overhead too in maintenance terms, and cross distro support. 

In the above  scenario, QubesOS already deals with launching the proper update script (apt, dnf) for TemplateVMs from the widget. I do not see what would be the difference here.

 On the solution side, I see 
- Fedora is already QubesOS documented for Fedora to pass required FS for in place upgrade doc (easy fix with dom0 provided overlay)
- Debian is not ([while documentation states how from configuration file](https://unix.stackexchange.com/questions/261165/change-cache-location-path-for-apt))
- Ubuntu would be the same
- Gentoo would be the same.

While upgrading from QubesOS update widget is already taking care of the needed backend calls, calling the proper TemplateVM generic script spawning the upgrade calls. The additional desired magic would still happen transparently while modifying the calls: the packages previously downloaded would be available in package cache for the all other cloned TemplateVMs, not being downloaded multiple times unnecessarily (which is the objective of this unfixed issue).
> The benefits of an existing caching proxy are that these issues have already been resolved.

Agreed that transparent proxy would be a magical solution that we would have loved, while you clearly stated [here](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-502108249) that this hack was not filling the need and would introduce other risks, while previous implementation is [just not working anymore ](https://github.com/unman/notes/tree/master/config/cacher).

Point being: [This was planned)](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-602097608) to be funded work undel NlNet's PET0 grant for the [Accessible Security project](https://nlnet.nl/project/AccessibleSecurity/). But this is [stalled since March 2019](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-596087003) without actual resolution being planned.

@andrewdavidwong we need feasibility assessment and amount of work needed prior of agreeing on payment, whic depends on the approach taken.

**Comment by unman** [comment]
On Fri, Sep 04, 2020 at 12:53:44PM -0700, tlaurion wrote:
> @unman : thanks for your reply.
> > Limiting to 1 VM template at a time? *very* sub-optimal. 
> 
> This is how Qubes Updates widget's launched upgrade process currently works: one TemplateVM being upgraded at a time.

I know and it's sub-optimal. The proposed solution enforces this.
Parallel upgrade using a caching proxy and you wont want to go back.

> > There's a significant overhead too in maintenance terms, and cross distro support. 
> 
> In the above  scenario, QubesOS already deals with launching the proper update script (apt, dnf) for TemplateVMs from the widget. I do not see what would be the difference here.

The widget does this by a salt call to pkg.uptodate. Customising the
local cache use would be different. (Not necessarily difficult, but
different.)

> 
>  On the solution side, I see 
> - Fedora is already QubesOS documented for Fedora to pass required FS for in place upgrade doc (easy fix with dom0 provided overlay)
> - Debian is not ([while documentation states how from configuration file](https://unix.stackexchange.com/questions/261165/change-cache-location-path-for-apt))
> - Ubuntu would be the same
> - Gentoo would be the same.
> 
> While upgrading from QubesOS update widget is already taking care of the needed backend calls, calling the proper TemplateVM generic script spawning the upgrade calls. The additional desired magic would still happen transparently while modifying the calls: the packages previously downloaded would be available in package cache for the all other cloned TemplateVMs, not being downloaded multiple times unnecessarily (which is the objective of this unfixed issue).
> > The benefits of an existing caching proxy are that these issues have already been resolved.
> 
> Agreed that transparent proxy would be a magical solution that we would have loved, while you clearly stated [here](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-502108249) that this hack was not filling the need and would introduce other risks, while previous implementation is [just not working anymore ](https://github.com/unman/notes/tree/master/config/cacher).
> 

I did not say that apt-cacher-ng didnt fill the need - it
does.
Nor did I say it introduced other risks. I have no idea what you mean.

And I dont know what you mean by a previous implementation "not working
anymore" - it works fine for almost all use cases, including Fedora. (My
notes are just that, and havent been updated for a while.)

> Point being: [This was planned)](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-602097608) to be funded work undel NlNet's PET0 grant for the [Accessible Security project](https://nlnet.nl/project/AccessibleSecurity/). But this is [stalled since March 2019](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-596087003) without actual resolution being planned.

What has been stalled since 2019? Agreement on a caching solution?

**Comment by tlaurion** [comment]
>>Agreed that transparent proxy would be a magical solution that we would have loved, while you clearly stated [here]([#1957 (comment)](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-502108249)) that this hack was not filling the need and would introduce other risks, while previous implementation is [just not working anymore ](https://github.com/unman/notes/tree/master/config/cacher).

> I did not say that apt-cacher-ng didnt fill the need - it does. Nor did I say it introduced other risks. I have no idea what you mean. And I dont know what you mean by a previous implementation "not working anymore" - it works fine for almost all use cases, including Fedora. (My notes are just that, and havent been updated for a while.)

Sorry, I didn't differentiate properly what is working (your apt-cacher-ng) while considered a hack, and what wasn't working anymore (@rustybird ).

In my understanding of apt-cacher-ng implementation, it would require repositories [to be modified](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-602110988), would complicate new user's understanding of how to properly add new repositories, which is not transparent. My critic is [here](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-602267843).

>> Point being: [This was planned)]([#1957 (comment)](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-602097608)) to be funded work undel NlNet's PET0 grant for the [Accessible Security project](https://nlnet.nl/project/AccessibleSecurity/). But this is [stalled since March 2019]([#1957 (comment)](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-596087003)) without actual resolution being planned.
> What has been stalled since 2019? Agreement on a caching solution?
Yes. So that we go forward in the realization of a solution that doesn't complexify user experience even more.

> Parallel upgrade using a caching proxy and you wont want to go back.
I wish. I do not see really well how racing condition between TemplateVMs download of packages can be mitigated if not launched in some ordered way (not totally parallelized).

@unman : What do you think about this [comment](url)? How would we make this transparent to the user who tried to add for example [Signal](https://signal.org/download/linux/) while not considering him a geek? This is why I would prefer a file cache. I do not think having the suboptimal, sequential upgrade of templates to be real problem if the next invocation of the upgrade actually installs the packages from the download cache, if that approach is totally transparent to the user and opposed to the apt-cacher-ng approach.

**Comment by unman** [comment]
On Sat, Sep 05, 2020 at 09:40:59AM -0700, tlaurion wrote:
> >>Agreed that transparent proxy would be a magical solution that we would have loved, while you clearly stated [here]([#1957 (comment)](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-502108249)) that this hack was not filling the need and would introduce other risks, while previous implementation is [just not working anymore ](https://github.com/unman/notes/tree/master/config/cacher).
> 
> > I did not say that apt-cacher-ng didnt fill the need - it does. Nor did I say it introduced other risks. I have no idea what you mean. And I dont know what you mean by a previous implementation "not working anymore" - it works fine for almost all use cases, including Fedora. (My notes are just that, and havent been updated for a while.)
> 
> Sorry, I didn't differentiate properly what is working (your apt-cacher-ng) while considered a hack, and what wasn't working anymore (@rustybird ).
> 

To be clear - the use of apt-cacher-ng isn't a hack - its a straight
implementation.

> In my understanding of apt-cacher-ng implementation, it would require repositories [to be modified](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-602110988), would complicate new user's understanding of how to properly add new repositories, which is not transparent. My critic is [here](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-602267843).

Yes, it is necessary to implement something to deal with TLS, and the
recommended method is to use `http://HTTPS///` in the repository
definitions.
This can be readily implemented either on template build or by salt
call on update. It can be absolutely transparent to users.
It's normal to also allow TLS pass through uncached.

> >> Point being: [This was planned)]([#1957 (comment)](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-602097608)) to be funded work undel NlNet's PET0 grant for the [Accessible Security project](https://nlnet.nl/project/AccessibleSecurity/). But this is [stalled since March 2019]([#1957 (comment)](https://github.com/QubesOS/qubes-issues/issues/1957#issuecomment-596087003)) without actual resolution being planned.
> > What has been stalled since 2019? Agreement on a caching solution?
> Yes. So that we go forward in the realization of a solution that doesn't complexify user experience even more.
> 
> > Parallel upgrade using a caching proxy and you wont want to go back.
> I wish. I do not see really well how racing condition between TemplateVMs download of packages can be mitigated if not launched in some ordered way (not totally parallelized).

Well, using (e.g) parallel just works because (I assume) templates fire
up at different rates. 

> 
> @unman : What do you think about this [comment](url)? How would we make this transparent to the user who tried to add for example [Signal](https://signal.org/download/linux/) while not considering him a geek? This is why I would prefer a file cache. I do not think having the suboptimal, sequential upgrade of templates to be real problem if the next invocation of the upgrade actually installs the packages from the download cache, if that approach is totally transparent to the user and opposed to the apt-cacher-ng approach.
> 

The user adds a repository as they will - if it is `https://` then it wont
be cached on first use.
The repository definition can be updated by the Qubes Updater, and an
explanatory note added to the relevant source definition file.
This is as transparent as can be.

To be clear, the caching qube idea isn't a bad one. I wouldn't use it
for obvious reasons.
Using a tool designed for the job seems better.

**Comment by marmarek** [comment]
I like the apt-cacher-ng approach in general. It sounds much cleaner on the architecture level than intercepting connections transparently, which can work only for non-TLS connections. Currently I see two issues with it, hopefully easy to solve:
 - requirement to modify repository config (different URL, especially for TLS)
 - risk of caching broken updates (normally when dnf downloads the package and it doesn't match the checksum, it retries from a different mirror - I guess with apt-cacher-ng it would simply re-download the same thing from the cache then...)

As for the repository config modification, it should be possible to do it in salt formula used for the update ([update.qubes-vm](https://github.com/QubesOS/qubes-mgmt-salt-dom0-update/blob/master/update/qubes-vm.sls)). But there is still and issue with updating the config by package manager then - if we modify the file, any updates to the config shipped via deb/rpm will need to be applied manually (using .rpmnew or .dpkg-new files). This doesn't sound like a big issue, but still something to worry about - if we ignore it, we may risk using outdated repository configs. Perhaps we should have some check for that and issue a warning to the user like "manual config file update is required"?

BTW, the option with sharing cache as a disk image is very risky. First, mounting a filesystem is a complex thing - one template can try to attack another template by planting malicious filesystem metadata for example. But then, I would assume that package managers consider local cache as trusted (things already verified before saving there) and may be less careful about data loaded from there. Even if packages itself may have signatures verified again, some other cached files (like repository metadata) may not.

**Comment by tasket** [comment]
> To be clear, the caching qube idea isn't a bad one. I wouldn't use it for obvious reasons.

Biggest problem with a caching or mirroring qube seems to be race conditions. But the update process is supposed to be orchestrated from dom0 anyway, so in my mind its not a stand out issue.

Once an updateVM has updated its cache, any number of templates can access the updated repo in parallel. This makes sense because if there are many templates to be updated, then waiting for the updateVM does not have a large cost proportionally and overall it helps performance; But if there are few templates then its a wash and the cached repo could be enabled or disabled with little difference.

OTOH, I understand the apt-cacher-ng approach to use a dynamic cache. If that's true then that raises a question about efficiency when users do not want all of their templates to update during every update cycle.

> BTW, the option with sharing cache as a disk image is very risky. First, mounting a filesystem is a complex thing - one template can try to attack another template by planting malicious filesystem metadata for example.

Probably not a concern if the repo is a separate volume attached in read-only mode (this theoretically allows it to be accessed by many templates simultaneously). However, I do think there is some risk from corruption of the filesystem itself – a different kind of metadata. If this isn't safe enough then other intra-system means are available.

> Even if packages itself may have signatures verified again, some other cached files (like repository metadata) may not.

That old chestnut: https://github.com/QubesOS/qubes-issues/issues/1919#issuecomment-689886883 But I guess this means Debian is the safer choice for users.

**Comment by tasket** [comment]
An attractive option would be to take a dynamic caching system like _apt-cacher-ng_, and pre-populate it with a package inclusion list. That would keep an up-to-date copy of packages used by a typical desktop installation, and that would cover the lions' share of packages used by various templates.

**Comment by tasket** [comment]
I would like to think I could do it justice right now, but the salt codebase looks opaque to me (no reflection on unman, I just don't know salt). I'll revisit the issue in future to see if anyone else has picked up your offer or if I find someone who can do so.

**Comment by iamahuman** [comment]
Just FYI: `apt-cacher-ng` appears to have [orphaned](https://pagure.io/releng/issue/8917) since Fedora 32 due to [being unmaintained](https://bugzilla.redhat.com/show_bug.cgi?id=1735456),

**Comment by unman** [comment]
On Sun, Sep 20, 2020 at 06:24:37PM -0700, iamahuman wrote:
> Just FYI: `apt-cacher-ng` appears to have [orphaned](https://pagure.io/releng/issue/8917) since Fedora 32 due to [being unmaintained](https://bugzilla.redhat.com/show_bug.cgi?id=1735456),
> 

I've not considered the Fedora side - is it unmaintained on Fedora? Do
we have any Fedora maintainers on board?
It is, of course, good on Debian.

**Comment by iamahuman** [comment]
@unman The maintainer is unresponsive&mdash;no response to FTBFS in `apt-cacher-ng`.

Yes, `apt-cacher-ng` is native to Debian. I suppose being able to run the update cache on a Fedora VM as well would be desirable, though.

**Comment by unman** [comment]
On Mon, Sep 21, 2020 at 04:02:45PM +0000, iamahuman wrote:
> @unman The maintainer is unresponsive&mdash;no response to FTBFS in `apt-cacher-ng`.

So do we have a Fedora maintainer onboard?

**Comment by fepitre** [comment]
I can try to propose to resurrect it from Fedora side. I'm not sure to have the current bandwidth for but I could give it a try if it is worth to do it.

**Comment by fepitre** [comment]
Just to let you know that I've updated `apt-cacher-ng` and as it was orphaned for 8+ weeks, I've requested a re-review: https://bugzilla.redhat.com/show_bug.cgi?id=1916884. On the Fedora devel list there is already one user which is pretty enthusiast about this resurrection. You can already test it by using my COPR repository `fepitre/fedora`. I'm currently using it with success on Fedora 32 AppVM.

**Comment by unman** [comment]
On Sat, Jan 16, 2021 at 12:51:32AM -0800, Fr??d??ric Pierret wrote:
> Just to let you know that I've updated `apt-cacher-ng` and as it was orphaned for 8+ weeks, I've requested a re-review: https://bugzilla.redhat.com/show_bug.cgi?id=1916884. On the Fedora devel list there is already one user which is pretty enthusiast about this resurrection. You can already test it by using my COPR repository `fepitre/fedora`. I'm currently using it with success on Fedora 32 AppVM.
> 

When you say "with success" do you mean "with success for Fedora"?
Can you post your acng.conf file, because Fedora remains a PITA.
Are you rewriting the sources files in the templates?

**Comment by fepitre** [comment]
> When you say "with success" do you mean "with success for Fedora"? Can you post your acng.conf file, because Fedora remains a PITA. Are you rewriting the sources files in the templates?

Yes success with Fedora 32. I'm currently running the build provided on my COPR repository. I'm using the almost default conf: https://gist.github.com/fepitre/fd490e04fe92bd023f77f0e03984b05c and I'm having success on caching Debian repositories. FYI, I'm not using the `qubes-updates-cache`. I've simply setup `apt-cacher-ng` into an AppVM then `qvm-connect-tcp :aptcachervm:3142` from another Debian AppVM and use proxy setting to `localhost:3142`.

**Comment by unman** [comment]
On Sat, Jan 16, 2021 at 07:50:08AM -0800, Fr??d??ric Pierret wrote:
> > When you say "with success" do you mean "with success for Fedora"? Can you post your acng.conf file, because Fedora remains a PITA. Are you rewriting the sources files in the templates?
> 
> Yes success with Fedora 32. I'm currently running the build provided on my COPR repository. I'm using the almost default conf: https://gist.github.com/fepitre/fd490e04fe92bd023f77f0e03984b05c and I'm having success on caching Debian repositories. FYI, I'm not using the `qubes-updates-cache`. I've simply setup `apt-cacher-ng` into an AppVM then `qvm-connect-tcp :aptcachervm:3142` from another Debian AppVM and use proxy setting to `localhost:3142`.
> 

I was asking if you had success in caching Fedora packages.
But you are skipping caching for most repositories. The config is the
difficult part.

**Comment by SvenSemmler** [comment]
@unman how would I check if fedora packages are cached? I am using apt-cacher-ng in a debian based qube based on your notes and it appears to work. The download speed of the packages in the fedora templates and the presence of the mirror sites as subdirectories in /var/cache make me think it works, but how can I be sure? ... and would that be unexpected? The discussion in this issue makes me doubt.

**Comment by unman** [comment]
On Sun, Jan 17, 2021 at 05:18:14PM -0800, Sven Semmler wrote:
> @unman how would I check if fedora packages are cached? I am using apt-cacher-ng in a debian based qube based on your notes and it appears to work. The download speed of the packages in the fedora templates and the presence of the mirror sites as subdirectories in /var/cache make me think it works, but how can I be sure? ... and would that be unexpected? The discussion in this issue makes me doubt.
> 

That will depend on how the repositories are set in the yum.repos
definition, and the configuration.
If the config includes a general PassThroughPattern, then that content
will not be cached. Since the majority of fedora links are https in
default, without special treatment they will not be cached.
You can always confirm caching by examining the directories in
/var/cache/apt-cacher-ng with `du -sh`.
Also, if you use Remap, you've probably already discovered that the
default fedora definitions are wholly inadequate, and need
supplementing. That said, fedora caching seems to work reasonably well.

**Comment by SvenSemmler** [comment]
Thank you!

**Comment by tlaurion** [comment]
I think apt-cache as bundled per @unman deserves way more attention.

Even more since shaker project (which salts otherwise complex user cases) including this caching proxy.

@marmarek : @unman now releases the package under his own repo, alongside other spec files that deploys salt scripts and actually deploys them as post install scripts.

cacher is one if those packages.

As referred in a closed issue here, what is missing, without qubes integration, is for the qubes updater to reapply the wildcard so that repositories are cached when applying updates on cloned templates on next run.

https://github.com/unman/shaker/issues/5#issuecomment-1210155101

**Comment by ben-grande** [comment]
> I like the apt-cacher-ng approach in general. It sounds much cleaner on the architecture level than intercepting connections transparently, which can work only for non-TLS connections. Currently I see two issues with it, hopefully easy to solve:
> 
> * requirement to modify repository config (different URL, especially for TLS)
> * risk of caching broken updates (normally when dnf downloads the package and it doesn't match the checksum, it retries from a different mirror - I guess with apt-cacher-ng it would simply re-download the same thing from the cache then...)
> 
> As for the repository config modification, it should be possible to do it in salt formula used for the update ([update.qubes-vm](https://github.com/QubesOS/qubes-mgmt-salt-dom0-update/blob/master/update/qubes-vm.sls)). But there is still and issue with updating the config by package manager then - if we modify the file, any updates to the config shipped via deb/rpm will need to be applied manually (using .rpmnew or .dpkg-new files). This doesn't sound like a big issue, but still something to worry about - if we ignore it, we may risk using outdated repository configs. Perhaps we should have some check for that and issue a warning to the user like "manual config file update is required"?

How do you imagine this notification? Is this a package manager hook that runs at last and notifiy in DomU or the qube sends a feature request to Dom0?

**Comment by tlaurion** [comment]
Also resolved under @ben-grande [qusal repo](https://github.com/ben-grande/qusal), under [sys-cacher recipes](https://github.com/ben-grande/qusal/tree/main/salt/sys-cacher)

**Comment by tlaurion** [comment]
> If necessary, we can create a meta-issue about the general problem and add all these as sub-issues.

My understanding was this issue was the meta issue.

**Comment by unman** [comment]
> Both unman's cacher and qusal's sys-cacher resolve this issue, while not packaged nor consequently offered as an installer option.

I havent looked at qusal's offering, but it will be fundamentally the
same as cacher.
cacher is packaged, of course, and has been for some years. If by
"offered as an installer option" you mean taken with the official qubes
packages, and installed with a checkbox, this isnt going to work
straight off. Why? Because current installations (unlike stock
sys-qubes), require a working network connection.
It would be trivial to work round this by including the Debian package
and dependencies within the packaged cacher, so that it can be installed offline.
My preference is for cacher to be installed on a minimal templates, but
there is no reason why this should be a deal breaker.
If there were interest in this, it could be done. (Is apt-cacher-ng
available for fedora for that case where users misguidedly elect not to
install Debian templates?)

**Comment by rishi-jat** [comment]
I’m considering working on this as a contrib package.

Before starting, I want to confirm the expected direction:

- Is the goal here to formalize an existing approach (e.g., cacher / sys-cacher) into an official contrib package, or is a new implementation expected?

- For acceptance, is it sufficient to support one update path (e.g., Debian via apt-cacher-ng or Fedora via dnf proxy), or does it need to handle both cleanly from the start?

- Are there defined expectations around integration with the update proxy (tinyproxy) vs running as a separate caching service in the UpdateVM?

- What are the minimum acceptance criteria for this to be considered a valid contrib package (packaging, default config, documentation, security considerations)?

Also, could you clarify the current bounty amount and payout conditions for this issue?

If this aligns, I’ll start by evaluating existing implementations and propose a concrete packaging approach.

**Comment by ben-grande** [comment]
You have commented [only on bounty issues (4 so far)]( https://github.com/qubesos/qubes-issues/issues?q=is%3Aissue%20state%3Aopen%20sort%3Aupdated-desc%20involves%3Arishi-jat) and [I don't see any PR made to Qubes (0 so far)](https://github.com/search?q=org%3AQubesOS+involves%3Arishi-jat&type=pullrequests). I don't think we can accept that. If you are only interested in bounty hunting, we would spend time training you for one PR than when the bounty completes, there is a low chance you'd keep up with the project.

**Comment by rishi-jat** [comment]
> You have commented [only on bounty issues (4 so far)](https://github.com/qubesos/qubes-issues/issues?q=is%3Aissue%20state%3Aopen%20sort%3Aupdated-desc%20involves%3Arishi-jat) and [I don't see any PR made to Qubes (0 so far)](https://github.com/search?q=org%3AQubesOS+involves%3Arishi-jat&type=pullrequests). I don't think we can accept that. If you are only interested in bounty hunting, we would spend time training you for one PR than when the bounty completes, there is a low chance you'd keep up with the project.

Thanks for the feedback understood.

I’d like to contribute to Qubes and build a track record here. Could you suggest a good first issue or smaller task I should start with? I’ll focus on that and work toward a solid PR.

After that, I can come back to larger issues like this one.

**Comment by ben-grande** [comment]
See issues with label "good first issue" without assignee and no linked PR: https://github.com/qubesos/qubes-issues/issues?q=is%3Aissue%20state%3Aopen%20sort%3Aupdated-desc%20label%3A%22good%20first%20issue%22%20no%3Aassignee.
